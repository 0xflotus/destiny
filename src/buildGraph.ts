import fs from "fs";
import path from "path";
import { findEdges } from "./findEdges";
import { addEdge } from "./addEdge";
import { Graph, OldGraph } from "./Graph";
import { resolveExtensionAndIndex } from "./resolveExtensionAndIndex";
import { importToAbsolutePath } from "./importToAbsolutePath";

export function buildGraph(folderPath: string) {
  const graph: Graph = {};
  const oldGraph: OldGraph = {};
  const totalFiles: string[] = [];
  const recurse = (currentFolderPath: string) => {
    const files = fs.readdirSync(currentFolderPath);
    // console.log(currentFolderPath, files);
    for (const file of files) {
      if (file === ".git") {
        continue;
      }
      // console.log(file);
      const fullPath = path.resolve(path.join(currentFolderPath, file));
      const start = path.relative(folderPath, fullPath);
      if (!(start in oldGraph)) {
        oldGraph[start] = { oldLocation: fullPath, imports: [] };
      }
      // check if it's a file if it has an extension
      if (fs.lstatSync(fullPath).isFile()) {
        findEdges(fullPath).forEach(edge => {
          const pathWithExtension = resolveExtensionAndIndex(
            importToAbsolutePath(edge[0], edge[1])
          );

          let end;
          if (!pathWithExtension) {
            console.log("Could not resolve import: ", edge);
            return;
          } else {
            end = path.relative(folderPath, pathWithExtension);
          }

          addEdge([start, end], graph);

          oldGraph[start].imports.push({
            text: edge[1],
            resolved: end
          });
        });
      } else {
        // console.log("recurse: ", fullPath);
        recurse(fullPath);
      }
      totalFiles.push(file);
    }
  };
  recurse(folderPath);
  return { graph, files: totalFiles, oldGraph };
}
