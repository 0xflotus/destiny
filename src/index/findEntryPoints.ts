import { Graph } from "./shared/Graph";

export function findEntryPoints(graph: Graph) {
  const importedFiles = new Set(Object.values(graph).flat());
  const possibleEntryPoints = Object.keys(graph);
  const entryPoints = possibleEntryPoints.filter(
    file => !importedFiles.has(file)
  );

  if (entryPoints.length) {
    return entryPoints;
  }

  const levelMap: Record<number, string[]> = {};

  possibleEntryPoints.forEach(k => {
    const n = k.split("/").length;

    if (!(n in levelMap)) {
      levelMap[n] = [];
    }

    levelMap[n].push(k);
  });

  for (let i = 1; i < 10; i++) {
    if (i in levelMap) {
      return levelMap[i];
    }
  }

  return [];
}
