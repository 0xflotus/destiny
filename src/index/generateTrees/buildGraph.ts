import path from "path";

import logger from "../../shared/logger";
import { Graph, OldGraph } from "./shared/Graph";
import { customResolve } from "../shared/customResolve";
import { findImports } from "../shared/findImports";
import { findSharedParent } from "./shared/findSharedParent";

const isFilePathIgnored = (filePath: string) => {
  const ignoreList = [/^\.git/];

  return ignoreList.some(re => re.test(filePath));
};

/** Build graph of all file paths and their own imports. */
export function buildGraph(filePaths: string[]) {
  const parentFolder = findSharedParent(filePaths);
  const graph: Graph = {};
  const oldGraph: OldGraph = {};
  const totalFiles: string[] = [];
  let numForwardSlashes = 0;
  let numBackSlashes = 0;

  for (let filePath of filePaths) {
    if (isFilePathIgnored(filePath)) continue;

    filePath = path.resolve(filePath);

    const start = path.relative(parentFolder, filePath);
    if (oldGraph.start == null) {
      oldGraph[start] = {
        oldLocation: filePath,
        imports: [],
      };
    }

    totalFiles.push(start);
    findImports(filePath).forEach(importPath => {
      if (importPath.includes("/")) {
        numForwardSlashes++;
      } else if (importPath.includes("\\")) {
        numBackSlashes++;
      }

      const pathWithExtension = customResolve(
        importPath,
        path.dirname(filePath)
      );

      if (pathWithExtension == null) {
        logger.error(`Cannot find import ${importPath}`);
        return;
      }

      const end = path.relative(parentFolder, pathWithExtension);

      if (!Array.isArray(graph[start])) {
        graph[start] = [];
      }
      if (!graph[start].includes(end)) {
        graph[start].push(end);
      }

      oldGraph[start].imports.push({
        text: importPath,
        resolved: end,
      });
    });
  }

  return {
    files: totalFiles,
    graph,
    oldGraph,
    parentFolder,
    useForwardSlash: numForwardSlashes >= numBackSlashes,
  };
}
