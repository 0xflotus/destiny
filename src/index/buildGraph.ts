import fs from "fs";
import path from "path";
import { findEdges } from "./buildGraph/findEdges";
import { addEdge } from "./buildGraph/addEdge";
import { Graph, OldGraph } from "./shared/Graph";
import { resolveExtensionAndIndex } from "./buildGraph/resolveExtensionAndIndex";
import { importToAbsolutePath } from "./buildGraph/importToAbsolutePath";

const isTestFile = (f: string) => /\.test\.|\.spec\./.test(f);

export function buildGraph(folderPath: string) {
  const graph: Graph = {};
  const oldGraph: OldGraph = {};
  const totalFiles: string[] = [];
  let numForwardSlashes = 0;
  let numBackSlashes = 0;
  const recurse = (currentFolderPath: string) => {
    const files = fs.readdirSync(currentFolderPath);
    // console.log(currentFolderPath, files);
    for (const file of files) {
      if (file === ".git") {
        continue;
      }
      // console.log(file);
      const fullPath = path.resolve(path.join(currentFolderPath, file));
      const start = path.relative(folderPath, fullPath);
      if (!(start in oldGraph)) {
        oldGraph[start] = { oldLocation: fullPath, imports: [] };
      }
      // check if it's a file if it has an extension
      if (fs.lstatSync(fullPath).isFile()) {
        totalFiles.push(start);
        findEdges(fullPath).forEach(edge => {
          if (edge[1].includes("/")) {
            numForwardSlashes++;
          } else if (edge[1].includes("\\")) {
            numBackSlashes++;
          }

          const pathWithExtension = resolveExtensionAndIndex(
            importToAbsolutePath(edge[0], edge[1])
          );

          let end;
          if (!pathWithExtension) {
            console.log("Could not resolve import: ", edge);
            return;
          } else {
            end = path.relative(folderPath, pathWithExtension);
          }

          // flip edge if the file is a .test.js or .spec.js file
          const isTestOrSpec = isTestFile(start);
          addEdge(isTestOrSpec ? [end, start] : [start, end], graph);

          oldGraph[isTestOrSpec ? end : start].imports.push({
            text: isTestOrSpec ? edge[0] : edge[1],
            resolved: end,
            reversed: isTestOrSpec,
          });
        });
      } else {
        recurse(fullPath);
      }
    }
  };
  recurse(folderPath);
  return {
    graph,
    files: totalFiles,
    oldGraph,
    useForwardSlash: numForwardSlashes >= numBackSlashes ? true : false,
  };
}
