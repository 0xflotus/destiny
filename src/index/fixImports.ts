import { readFileSync, writeFileSync } from "fs-extra";
import resolve from "resolve";
import { findEdges } from "./formatFileStructure/buildGraph/findEdges";
import path from "path";
import { makeImportPath } from "./formatFileStructure/syncFileSystem/fixImports/makeImportPath";

export const fixImports = (
  files: string[],
  parentFolder: string,
  newStructure: Record<string, string>,
  useForwardSlashes: boolean
) => {
  for (const file of files) {
    const imports = findEdges(file);
    if (!imports.length) {
      continue;
    }

    const basedir = path.dirname(file);

    const ogText = readFileSync(file).toString();
    // deep copy
    let newText = ogText.repeat(1);
    for (const imp of imports) {
      const absPath = resolve.sync(imp[1], { basedir });
      const pathInGraph = path.relative(parentFolder, absPath);
      const newPath = newStructure[pathInGraph];
      if (!newPath) {
        continue;
      }

      newText = newText.replace(
        imp[1],
        makeImportPath(file, path.resolve(newPath), useForwardSlashes)
      );
    }
    if (newText !== ogText) {
      writeFileSync(file, newText);
    }
  }
};
