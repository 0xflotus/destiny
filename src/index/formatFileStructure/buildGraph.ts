import fs from "fs";
import path from "path";
import { findEdges } from "./buildGraph/findEdges";
import { addEdge } from "./buildGraph/addEdge";
import { Graph, OldGraph } from "./shared/Graph";
import { resolveExtensionAndIndex } from "./buildGraph/resolveExtensionAndIndex";
import { importToAbsolutePath } from "./buildGraph/importToAbsolutePath";

export function buildGraph(
  folderPath: string,
  throwIfCannotBeResolved = false
) {
  const graph: Graph = {};
  const oldGraph: OldGraph = {};
  const totalFiles: string[] = [];
  let numForwardSlashes = 0;
  let numBackSlashes = 0;
  const recurse = (currentFolderPath: string) => {
    const files = fs.readdirSync(currentFolderPath);
    // console.log(currentFolderPath, files);
    for (const file of files) {
      if (file === ".git") {
        continue;
      }
      // console.log(file);
      const fullPath = path.resolve(path.join(currentFolderPath, file));
      const start = path.relative(folderPath, fullPath);
      if (!(start in oldGraph)) {
        oldGraph[start] = {
          oldLocation: fullPath,
          imports: [],
        };
      }
      // check if it's a file if it has an extension
      if (fs.lstatSync(fullPath).isFile()) {
        totalFiles.push(start);
        findEdges(fullPath).forEach(edge => {
          if (edge[1].includes("/")) {
            numForwardSlashes++;
          } else if (edge[1].includes("\\")) {
            numBackSlashes++;
          }

          const pathWithExtension = resolveExtensionAndIndex(
            importToAbsolutePath(edge[0], edge[1])
          );

          let end;
          if (!pathWithExtension) {
            const msg = "Could not resolve import: " + edge;
            if (throwIfCannotBeResolved) {
              throw new Error(msg);
            } else {
              console.log(msg);
            }
            return;
          } else {
            end = path.relative(folderPath, pathWithExtension);
          }

          addEdge([start, end], graph);

          oldGraph[start].imports.push({
            text: edge[1],
            resolved: end,
          });
        });
      } else {
        recurse(fullPath);
      }
    }
  };
  recurse(folderPath);
  return {
    graph,
    files: totalFiles,
    oldGraph,
    useForwardSlash: numForwardSlashes >= numBackSlashes ? true : false,
  };
}
