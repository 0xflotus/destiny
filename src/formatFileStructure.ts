import { buildGraph } from "./index/buildGraph";
import { findEntryPoints } from "./index/findEntryPoints";
import { syncFileSystem } from "./index/syncFileSystem";
import { toFractalTree } from "./index/toFractalTree";
import { removeEmptyFolders } from "./index/removeEmptyFolders";
import { flatten } from "./shared/flatten";
export const formatFileStructure = async (start: string) => {
  const { graph, oldGraph, files, useForwardSlash } = buildGraph(start);
  const tree = toFractalTree(graph, findEntryPoints(graph));
  await syncFileSystem({
    originalGraph: oldGraph,
    newStructure: tree,
    destination: start,
    useForwardSlashes: useForwardSlash,
    startingFolder: start,
  });
  removeEmptyFolders(start);
  const usedFiles = new Set([
    ...Object.keys(graph),
    ...flatten(Object.values(graph)),
  ]);
  const unusedFiles: string[] = [];
  files.forEach(file => {
    if (!usedFiles.has(file)) {
      unusedFiles.push(file);
    }
  });
  if (unusedFiles.length) {
    console.log("unused files:");
    unusedFiles.forEach(f => console.log(f));
  }
};
