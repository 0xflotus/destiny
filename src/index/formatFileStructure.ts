import { buildGraph } from "./formatFileStructure/buildGraph";
import { findEntryPoints } from "./formatFileStructure/findEntryPoints";
import { moveFiles } from "./formatFileStructure/moveFiles";
import { toFractalTree } from "./formatFileStructure/toFractalTree";
import { removeEmptyFolders } from "./formatFileStructure/removeEmptyFolders";
import { fixImports } from "./formatFileStructure/fixImports";
import { RootOption } from "./shared/RootOption";

/*
  rootDirFiles: [
    ['pages/index.js', 'pages/register.js'],
    ['utils/search.js', 'utils/helper.js']
  ]
*/
export const formatFileStructure = async (
  rootDirFiles: string[][],
  filesToFixImports: string[]
) => {
  const unusedFiles: string[] = [];
  const rootOptions: RootOption[] = [];
  for (const startingFiles of rootDirFiles) {
    if (!startingFiles.length) {
      continue;
    }
    const { graph, files, useForwardSlash, parentFolder } = buildGraph(
      startingFiles
    );
    const tree = toFractalTree(graph, findEntryPoints(graph));
    rootOptions.push({
      tree,
      useForwardSlash,
      parentFolder,
    });
    const usedFiles = new Set([
      ...Object.keys(graph),
      ...Object.values(graph).flat(),
    ]);
    files.forEach(file => {
      if (!usedFiles.has(file)) {
        unusedFiles.push(file);
      }
    });
  }
  await fixImports(filesToFixImports, rootOptions);
  for (const { tree, parentFolder } of rootOptions) {
    await moveFiles(tree, parentFolder);
    removeEmptyFolders(parentFolder);
  }
  if (unusedFiles.length) {
    console.log("unused files:");
    unusedFiles.forEach(f => console.log(f));
  }
};
