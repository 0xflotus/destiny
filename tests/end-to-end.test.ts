import fs from "fs-extra";
import path from "path";
import { formatFileStructure } from "../src/index/formatFileStructure";
import { buildGraph } from "../src/index/formatFileStructure/buildGraph";
import treeDir from "tree-node-cli";

const tmpPath = path.join(__dirname, "tmp");

beforeAll(() => {
  fs.ensureDir(tmpPath);
});

afterAll(() => {
  fs.removeSync(tmpPath);
});

const treeDirWithContents = (dir: string) => {
  const files = fs.readdirSync(dir);
  let tree: Record<string, string> = {};
  for (const file of files) {
    const filePath = path.join(dir, file);
    if (fs.lstatSync(filePath).isDirectory()) {
      tree = {
        ...tree,
        ...treeDirWithContents(filePath),
      };
    } else {
      tree[path.relative(tmpPath, filePath)] = fs
        .readFileSync(filePath)
        .toString();
    }
  }

  return tree;
};

describe("end-to-end", () => {
  const fixturePath = path.join(__dirname, "fixtures");
  const testCases = fs.readdirSync(fixturePath);
  for (const ogTestCase of testCases) {
    const testCase =
      ogTestCase === "globals" ? path.join("globals", "src") : ogTestCase;
    it(testCase, async () => {
      const testCasePath = path.join(fixturePath, ogTestCase);
      fs.copySync(testCasePath, path.join(tmpPath, ogTestCase));
      const ogCopyPath = path.join(tmpPath, ogTestCase);
      const copyPath = path.join(tmpPath, testCase);
      await formatFileStructure(copyPath);
      // make sure no imports broke
      buildGraph(ogCopyPath, true);
      expect(treeDir(ogCopyPath)).toMatchSnapshot();
      const treeContents = treeDirWithContents(ogCopyPath);
      Object.keys(treeContents).forEach(k => {
        expect(treeContents[k]).toMatchSnapshot(k);
      });
    });
  }
});
