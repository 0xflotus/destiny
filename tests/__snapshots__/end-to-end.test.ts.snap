// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`end-to-end globals/src 1`] = `
"globals
├── global.js
└── src
    ├── index
    │   └── helper.js
    └── index.js"
`;

exports[`end-to-end globals/src: globals/global.js 1`] = `
"export const bob = 5;
"
`;

exports[`end-to-end globals/src: globals/src/index.js 1`] = `
"require(\\"./index/helper\\");
"
`;

exports[`end-to-end globals/src: globals/src/index/helper.js 1`] = `
"require(\\"../../global\\");
export const five = 5;
"
`;

exports[`end-to-end index-cycle 1`] = `
"index-cycle
├── index
│   ├── login
│   │   └── search.js
│   ├── login.js
│   ├── routes
│   │   └── home.js
│   └── routes.js
└── index.js"
`;

exports[`end-to-end index-cycle: index-cycle/index.js 1`] = `
"export { five } from \\"./index/routes\\";
export { seven } from \\"./index/login\\";
"
`;

exports[`end-to-end index-cycle: index-cycle/index/login.js 1`] = `
"import { six } from \\"./login/search\\";

export const seven = six + 1;
"
`;

exports[`end-to-end index-cycle: index-cycle/index/login/search.js 1`] = `
"import { five } from \\"../../index\\";

export const six = five + 1;
"
`;

exports[`end-to-end index-cycle: index-cycle/index/routes.js 1`] = `
"export * from \\"./routes/home\\";
"
`;

exports[`end-to-end index-cycle: index-cycle/index/routes/home.js 1`] = `
"export const five = 5;
"
`;

exports[`end-to-end sharing 1`] = `
"sharing
├── index
│   ├── footer.js
│   ├── header.js
│   └── shared
│       └── helper.js
└── index.js"
`;

exports[`end-to-end sharing: sharing/index.js 1`] = `
"export * from \\"./index/footer\\";
export * from \\"./index/header\\";
"
`;

exports[`end-to-end sharing: sharing/index/footer.js 1`] = `
"import { five } from \\"./shared/helper\\";

export const seven = five + 2;
"
`;

exports[`end-to-end sharing: sharing/index/header.js 1`] = `
"import { five } from \\"./shared/helper\\";
export const six = five + 1;
"
`;

exports[`end-to-end sharing: sharing/index/shared/helper.js 1`] = `
"export const five = 5;
"
`;

exports[`end-to-end simple 1`] = `
"simple
├── index
│   ├── routes
│   │   └── home.js
│   └── routes.js
└── index.js"
`;

exports[`end-to-end simple: simple/index.js 1`] = `
"import { five } from \\"./index/routes\\";
"
`;

exports[`end-to-end simple: simple/index/routes.js 1`] = `
"export * from \\"./routes/home\\";
"
`;

exports[`end-to-end simple: simple/index/routes/home.js 1`] = `
"export const five = 5;
"
`;

exports[`end-to-end spec-files 1`] = `
"spec-files
├── index
│   ├── level1
│   │   ├── level2.js
│   │   └── level2.spec.js
│   ├── level1.js
│   └── level1.spec.js
├── index.js
└── index.spec.js"
`;

exports[`end-to-end spec-files: spec-files/index.js 1`] = `
"require(\\"./index/level1\\");
"
`;

exports[`end-to-end spec-files: spec-files/index.spec.js 1`] = `
"require(\\"./index\\");
"
`;

exports[`end-to-end spec-files: spec-files/index/level1.js 1`] = `
"require(\\"./level1/level2\\");
"
`;

exports[`end-to-end spec-files: spec-files/index/level1.spec.js 1`] = `
"require(\\"./level1\\");
"
`;

exports[`end-to-end spec-files: spec-files/index/level1/level2.js 1`] = `""`;

exports[`end-to-end spec-files: spec-files/index/level1/level2.spec.js 1`] = `
"require(\\"./level2\\");
"
`;
